src/server/tweetFinder.js

```diff
@@ -1,63 +1,85 @@
+        res.end(JSON.stringify(data));
       },
-    })
-    .then((tweetRecords) => {
-      const result = [];
-
-      const promises = tweetRecords.map((tweetRecord) => {
-        const rid = '' + tweetRecord['@rid'];
-
-        return db.query(
-          `SELECT FROM (TRAVERSE in(TWEETED) FROM (SELECT FROM ${rid})) WHERE @class = "Tweeter"`
-          )
-          .then((tweeterRecords) => {
-            const tweeterRecord = tweeterRecords[0];
-            const toReturn = {
-              'tweet':
-                flattenImmutableObject(
-                  TweetBuilder()
-                    .id(tweetRecord.id)
-                  .content(tweetRecord.content)
-                  .date(tweetRecord.date.toISOString())
-                  .likes(tweetRecord.likes)
-                  .retweets(tweeterRecord.retweets)
-                  .build()),
-              'tweeter':
-                flattenImmutableObject(
-                  TweeterBuilder()
-                    .id(tweeterRecord.id)
-                  .name(tweeterRecord.name)
-                  .handle(tweeterRecord.handle)
-                  .build()),
-            };
-
-            result.push(toReturn);
-          });
-      });
-
-      Promise.all(promises)
-        .then(() => {
-          response.end(
-            JSON.stringify(
-              result));
+
+      (rejection) => {
+        res.writeHead(500, { 'Content-Type': 'application/json' });
+        res.end('An unexpected internal error occurred.');
+        console.warn(`Unable to search for query '${query}'`, rejection);
+      }
+    );
+};
+
+const searchAndCollateResults = (query) => {
+  return searchDatabase(query)
+    .then((data) => {
+      return {
+        'data': {
+          'count': data.length,
+          'records': getTweetsAsResults(data),
+        },
+      };
+    });
+};
+
+const searchDatabase = (query, alreadyAttemptedRefresh = false) => {
+  const tweetSelection = 'SELECT FROM tweet WHERE content LUCENE :query ORDER BY date DESC LIMIT 300';
+
+  return chainPromises(() => {
+    return db.query(tweetSelection, { 'params': { 'query': `${query}~` } });
+  }).then(
+    (tweetRecords) => {
+      const shouldRequeryTwitter = !alreadyAttemptedRefresh && tweetRecords.length <= 20;
+      if (shouldRequeryTwitter) {
+        return refreshFromTwitter(query);
+
+      } else {
+        return tweetRecords.map((tweetRecord) => {
+          return flattenImmutableObject(buildTweetFromDatabaseRecord(tweetRecord));
         });
+      }
+    },
+
+    (rejection) => {
+      console.warn('Major error querying the database.', rejection);
+    }
+  );
+};
 
-    })
-    .error((error) => {
-      console.error('Major issues abound.');
-      response.end('Unable to connect to database.');
+const refreshFromTwitter = (query) => {
+  return searchAndSaveFromTwitter(query)
+    .then(() => {
+      return searchDatabase(query, true);
     });
 };
+
+const buildTweetFromDatabaseRecord = (record) => {

```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: ['buildTweetFromDatabaseRecord' was used before it was defined.](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2180581610/issues/source?bid=3148782&fileBranchId=3164498#l73)